mtt-management/src/mtt_types.rs

```diff
@@ -0,0 +1,446 @@
+    MttConfidential4K = 1,
+    /// The L1 entry corresponding to the address is invalid.
+    /// This should be treated as an invalid operational state.
+    MttInvalidL1,
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+/// API-related enums for non-confidential and confidential types
+pub enum MttMemoryType {
+    /// A non-confidential memory range
+    NonConfidential = 0,
+    /// A confidential memory range
+    Confidential = 1,
+}
+
+// Bits 33:0
+const MTT_L2_INFO_MASK: u64 = 0x0003_ffff_ffff;
+
+use MttL2EntryType::*;
+impl MttL2EntryType {
+    fn from_raw(value: u64) -> MttL2EntryType {
+        // Bits 63:35 must be zero
+        if (value & 0xffff_fff0_0000_0000) != 0 {
+            return MttInvalidL2;
+        }
+
+        // Bits 35:34 are the type
+        let entry_type = (value & 0x000c_0000_0000) >> 34;
+        match entry_type {
+            // INFO must be 0 for 1GB entries
+            0 if (value & MTT_L2_INFO_MASK) == 0 => MttNonConfidential1G,
+            1 if (value & MTT_L2_INFO_MASK) == 0 => MttConfidential1G,
+            2 => MttL1Dir(value & MTT_L2_INFO_MASK),
+            3 => Mtt2MBPages(value & MTT_L2_INFO_MASK),
+            _ => MttInvalidL2,
+        }
+    }
+
+    fn raw(&self) -> u64 {
+        match *self {
+            MttNonConfidential1G => 0,
+            MttConfidential1G => 1u64 << 34,
+            MttL1Dir(v) => (2u64 << 34) | v,
+            Mtt2MBPages(v) => (3u64 << 34) | v,
+            _ => unreachable!(),
+        }
+    }
+
+    fn confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0xffff_ffff)
+    }
+
+    fn non_confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0)
+    }
+
+    fn mtt_l1_dir(mtt_l1_page: SupervisorPageAddr) -> MttL2EntryType {
+        MttL1Dir(mtt_l1_page.bits() >> 12)
+    }
+}
+
+use MttL1EntryType::*;
+impl MttL1EntryType {
+    fn from_raw(value: u64) -> MttL1EntryType {
+        match value {
+            0 => MttNonConfidential4K,
+            1 => MttConfidential4K,
+            _ => MttInvalidL1,
+        }
+    }
+}
+
+/// Errors returned by the mtt_types crate
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The 4K memory-region couldn't be converted because there was no existing
+    /// L1 entry for it, and the caller didn't pass a 4K page for a new L1 entry.
+    L1TableNeeded,
+}
+
+/// Result from mtt_crate APIs
+pub type Result<T> = core::result::Result<T, Error>;
+
+fn get_l2_entry(phys_addr: u64) -> MttL2EntryType {
+    // Bits 45:26, shifted right by 26, then 3-bit left shift to account for the 8-byte entry
+    let mtt_index = (phys_addr & 0x3fff_fc00_0000) >> 23;
+    let mtt_entry_addr = (get_mtt_base() + mtt_index) as *const u64;
+    // Safety: The MTT is a 8MB contiguous region allocated by trusted platform-FW. The TSM is
+    // the only entity responsible for maintaining it, and we are using a 23-bit index into it
+    let value = unsafe { *mtt_entry_addr };
```

Comment:
rust newb question - isn't there a safe way to refer to this 8MB region via an object that owns that 8MB space?

Reply:
Ultimately, we'll need an unsafe block somewhere to actually write to the 8MB region.